Automattic/newspack-plugin
# Feat/nrh xdomain tracking

### All Submissions:






### Changes proposed in this Pull Request:

Adds [Google Analytics Cross Domain tracking](https://developers.google.com/gtagjs/devguide/linker.) for sites using News Revenue Hub checkout. For the feature to work, the Donate block needs to submit directly to the NRH checkout site — the `decorate_forms` feature won't work with the redirect approach currently on `master`. Because of this problem, the biggest part of this branch is a rewrite of the Donate block rendering function for sites with News Revenue Hub checkout. This also gets a bit complicated because of how the form inputs work. In the new implementation there's a hidden `amount` field which should contain the current amount accurately following a variety of form interactions, including:

- Initial values that reflect the default tab
- Clicking on the three price buttons
- Switching tabs
- Manually setting the amount after clicking the Other button

I noticed that the `https://checkout.fundjournalism.org/donateform` endpoint now seems to always redirect to `https://checkout.fundjournalism.org/memberform`, so I skipped implementing the logic to switch between the two endpoints (original implementation [here](https://github.com/Automattic/newspack-plugin/blob/master/includes/class-nrh.php#L60-L72)).

Test with https://github.com/Automattic/newspack-blocks/pull/671.

### How to test the changes in this Pull Request:

1. In the Reader Revenue wizard switch platform to News Revenue Hub. Use an Organization ID and Salesforce Campaign ID from a production site.
2. Google Analytics will need to be set up for cross-domain tracking to work.
3. Switch Newspack Blocks to `feat/attributes-in-donate-block-hook`
4. Create a few pages with the Donate block on them. Some should be AMP, some non-AMP. Some should have multiple Donate blocks on the same page. Use tiered and un-tiered, and experiment with different amounts and default tabs.
5. The easiest way to test that the hidden `amount` field is working is to change the type to `text` [here](https://github.com/Automattic/newspack-plugin/pull/810/files#diff-ae6055b360153ad3a025eaee30a5deade1801c95c3b03604c638a8372634118aR219) and [here](https://github.com/Automattic/newspack-plugin/pull/810/files#diff-ae6055b360153ad3a025eaee30a5deade1801c95c3b03604c638a8372634118aR318). This allows you to observe the value, and you can interact with the block in a variety of ways and verify the value is correct. Test in AMP and non-AMP, because the implementations are quite different.
6. Verify that the correct price and term is set when clicking through to the hub checkout page.
7. Verify the cross-domain tracking is working. There should be a `_ga` parameter appended to the URL that looks something like this: `_ga=2.257424207.531864944.1610926274-amp-V_WelaH4thIfJKrlpv_1HQ`
8. Bonus: add a button block to a page and link directly to an NRH checkout page. Verify the `_ga` parameter is appended to this URL too when clicked.
9. Verify none of this affects the behavior of the block on non-NRH sites.
10. Probably wise to verify the block works properly on sites that are still using the NRH checkout plugin and don't have NRH selected in the Reader Revenue wizard. Even though this is a configuration we should deprecate, we don't want to break donation on live sites.

### Other information:
